File: src/backend/bedrock-prompt-messager/services/bedrock_service.py
from datetime import datetime
import base64
import boto3
import json
import re
from core.config import BedrockConfig
from core import constants

CORS_HEADERS = {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "POST, OPTIONS",
    "Access-Control-Allow-Headers": "Content-Type,Authorization,X-Api-Key,X-Amz-Security-Token",
    "Content-Type": "application/json",
}

DATA_URL_PATTERN = re.compile(r'data\:([^;]+);base64,(.*)')

class BedrockService:
    def __init__(self):
        self.config = BedrockConfig()
        self.bedrock_runtime = boto3.client(
            service_name="bedrock-runtime", region_name=self.config.region
        )

    def call_admail_bedrock_prompt(self, input_letter, file_name):
        try:
            with open("system_prompt.txt", mode="r", encoding="utf-8") as prompt_file:
                system_prompt = prompt_file.read()
        except FileNotFoundError:
            return {"statusCode": 400, "body": constants.ERROR_SYSTEM_PROMPT_NOT_FOUND}

        try:
            mime, b64_bytes = DATA_URL_PATTERN.match(input_letter).groups()
        except AttributeError:
            return {"statusCode": 400, "body": "Invalid data url passed to bedrock service"}

        format = None
        if mime == 'application/pdf':
            format = 'pdf'
        elif mime == 'application/vnd.openxmlformats-officedocument.wordprocessingml.document':
            format = 'docx'
        elif mime == 'text/plain':
            format = 'txt'

        if not format:
            return {"statusCode": 400, "body": f"Unknown document format for mime type: {mime}"}

        user_prompt = "Analyze the following letter:"
        

        guardrail_assessment = self.bedrock_runtime.apply_guardrail(
            guardrailIdentifier=self.config.guardrail,
            guardrailVersion=self.config.guardrail_version,
            source="INPUT",
            content=[{"text": {"text": user_prompt}}],
        )

        messages = [
            { 
                'role': 'user', 
                'content': [
                    { 'text': user_prompt },
                    { 'document': { 
                        'format': format,
                        'name': 'the_letter',
                        'source': {
                            'bytes': base64.b64decode(b64_bytes)
                        }
                    }}
                ]
            }
        ]

        inference_config = {
            "temperature": self.config.temperature,
            "topP": self.config.top_p,
            "maxTokens": self.config.max_tokens,
        }

        response = self.bedrock_runtime.converse(
            modelId=self.config.model_id,
            system=[{"text": system_prompt}],
            messages=messages,
            inferenceConfig=inference_config,
            toolConfig=self._get_admail_tool_config(),
        )

        formatted_response = self.format_converse_response(response)
        api_gateway_response = {
            "statusCode": 200,
            "body": formatted_response,
            "headers": CORS_HEADERS,
        }
        self.log_prompt_details_to_s3(
            promptinput=user_prompt,
            promptoutput=api_gateway_response,
            guardrail_assessment=guardrail_assessment,
            filename=file_name,
        )
        return api_gateway_response

    def format_converse_response(self, response):
        response_message = response["output"]["message"]

        tool_use_block = next(
            (
                content
                for content in response_message["content"]
                if "toolUse" in content
            ),
            None,
        )

        if tool_use_block:
            tool_input = tool_use_block["toolUse"]["input"]
            formatted_content = json.dumps(tool_input, indent=4)
        else:
            formatted_content = response_message["content"][0]["text"]

        return formatted_content

    def _get_admail_tool_config(self):
        return {
            "tools": [
                {
                    "toolSpec": {
                        "name": constants.TOOL_NAME,
                        "description": constants.TOOL_DESCRIPTION,
                        "inputSchema": {
                            "json": {
                                "type": "object",
                                "properties": {
                                    "description": {
                                        "type": "string",
                                        "description": "Brief description of the letter or mailing content.",
                                    },
                                    "rating": {
                                        "type": "string",
                                        "description": "The eligibility rating for AdMail, as defined by ourselves, use with our prompt",
                                        "enum": [
                                            constants.RATING_BUSINESS,
                                            constants.RATING_UNSURE,
                                            constants.RATING_ADVERTISING,
                                        ],
                                    },
                                    "reason": {
                                        "type": "string",
                                        "description": "Bullet pointed explaination of letter eligibility for Admail",
                                    },
                                    "advice": {
                                        "type": "string",
                                        "description": "Actionable bullet points to convert the letter to Admail, if applicable.",
                                    },
                                },
                                "required": [
                                    "description",
                                    "rating",
                                    "reason",
                                    "advice",
                                ],
                            }
                        },
                    }
                }
            ],
            "toolChoice": {"tool": {"name": constants.TOOL_NAME}},
        }

    def log_prompt_details_to_s3(
        self, promptinput, promptoutput, guardrail_assessment, filename
    ):
        if not self.config.logging_s3_bucket or not self.config.logging_s3_key_prefix or not self.config.logging_s3_account_id:
            print(constants.ERROR_S3_LOGGING_NOT_CONFIGURED)
            return

        s3_client = boto3.client("s3")
        date_time_now = datetime.now().strftime("%d-%m-%Y_%H:%M:%S")
        s3_key = (
            f"{self.config.logging_s3_key_prefix}{date_time_now}|~{filename}|~.json"
        )

        log_data = {
            "prompt_input": promptinput,
            "prompt_output": promptoutput,
            "guardrail_assessment": guardrail_assessment,
            "model": self.config.model_id,
            "inference_parameters": {
                "temperature": self.config.temperature,
                "top_p": self.config.top_p,
                "max_tokens": self.config.max_tokens,
            },
            "date_time": date_time_now,
        }

        try:
            s3_client.put_object(
                Bucket=self.config.logging_s3_bucket,
                Key=s3_key,
                Body=json.dumps(log_data, indent=4),
                ContentType="application/json",
                ExpectedBucketOwner=self.config.logging_s3_account_id,
            )
        except Exception as e:
            print(f"Error logging to S3: {e}")

File: lambdas/bedrock-prompt-messager/src/bedrock-service.ts
import {
  ApplyGuardrailCommand,
  BedrockRuntimeClient,
  ContentBlock,
  ConverseCommand,
  type ConverseCommandOutput,
  DocumentFormat,
  type ToolConfiguration,
} from "@aws-sdk/client-bedrock-runtime";
import { PutObjectCommand, S3Client } from "@aws-sdk/client-s3";
import type { APIGatewayProxyResult } from "aws-lambda";

import type { BedrockConfig } from "src/config";
import {
  CORS_HEADERS,
  ERROR_MESSAGES,
  Rating,
  TOOL_DESCRIPTION,
  TOOL_NAME,
} from "src/constants";
import systemPromptExtended from "src/system_prompt_extended.txt";
import systemPrompt from "src/system_prompt.txt";

const DATA_URL_PATTERN = /^data:([^;]+);base64,([\S\s]+)$/;

const MIME_TO_FORMAT: Partial<Record<string, DocumentFormat>> = {
  "application/pdf": DocumentFormat.PDF,
  "application/vnd.openxmlformats-officedocument.wordprocessingml.document":
    DocumentFormat.DOCX,
  "text/plain": DocumentFormat.TXT,
};

export interface LogPromptDetailsInput {
  promptInput: string;
  promptOutput: APIGatewayProxyResult;
  guardrailAssessment: unknown;
  fileName: string | undefined;
}

const pad = (n: number): string => String(n).padStart(2, "0");

/** Formats a Date as DD-MM-YYYY_HH:MM:SS to match the Python logging key format. */
function formatDateTimeForKey(date: Date): string {
  return (
    `${pad(date.getDate())}-${pad(date.getMonth() + 1)}-${date.getFullYear()}` +
    `_${pad(date.getHours())}:${pad(date.getMinutes())}:${pad(date.getSeconds())}`
  );
}

export class BedrockService {
  constructor(
    private readonly config: BedrockConfig,
    private readonly bedrockClient: BedrockRuntimeClient,
    private readonly s3Client: S3Client,
  ) {}

  async callAdmailBedrockPrompt(
    inputLetter: string,
    fileName?: string,
  ): Promise<APIGatewayProxyResult> {
    const prompt = process.env.FEAT_EXTENDED_ADVICE
      ? systemPromptExtended
      : systemPrompt;

    const match = DATA_URL_PATTERN.exec(inputLetter.trim());
    if (!match) {
      return { statusCode: 400, body: ERROR_MESSAGES.INVALID_DATA_URL };
    }

    const [, mime, b64] = match;
    // eslint-disable-next-line security/detect-object-injection
    const format = MIME_TO_FORMAT[mime];
    if (!format) {
      return {
        statusCode: 400,
        body: ERROR_MESSAGES.UNKNOWN_MIME_TYPE(mime),
      };
    }

    const userPrompt = "Analyze the following letter:";

    const guardrailResponse = await this.bedrockClient.send(
      new ApplyGuardrailCommand({
        guardrailIdentifier: this.config.guardrail,
        guardrailVersion: this.config.guardrailVersion,
        source: "INPUT",
        content: [{ text: { text: userPrompt } }],
      }),
    );

    const converseResponse = await this.bedrockClient.send(
      new ConverseCommand({
        modelId: this.config.modelId,
        system: [{ text: prompt }],
        messages: [
          {
            role: "user",
            content: [
              { text: userPrompt },
              {
                document: {
                  format,
                  name: "the_letter",
                  source: { bytes: Buffer.from(b64, "base64") },
                },
              },
            ],
          },
        ],
        inferenceConfig: {
          temperature: this.config.temperature,
          topP: this.config.topP,
          maxTokens: this.config.maxTokens,
        },
        toolConfig: BedrockService.getAdmailToolConfig(),
      }),
    );

    const formattedBody =
      BedrockService.formatConverseResponse(converseResponse);
    const apiGatewayResponse: APIGatewayProxyResult = {
      statusCode: 200,
      body: formattedBody,
      headers: CORS_HEADERS,
    };

    await this.logPromptDetailsToS3({
      promptInput: userPrompt,
      promptOutput: apiGatewayResponse,
      guardrailAssessment: guardrailResponse,
      fileName,
    });

    return apiGatewayResponse;
  }

  static formatConverseResponse(response: ConverseCommandOutput): string {
    const content = response.output?.message?.content ?? [];

    const toolUseBlock = content.find(
      (block): block is ContentBlock.ToolUseMember =>
        "toolUse" in block && block.toolUse !== undefined,
    );

    if (toolUseBlock) {
      return JSON.stringify(toolUseBlock.toolUse.input, null, 4);
    }

    const textBlock = content[0];
    if (textBlock && "text" in textBlock) {
      return textBlock.text ?? "";
    }

    return "";
  }

  static getAdmailToolConfig(): ToolConfiguration {
    return {
      tools: [
        {
          toolSpec: {
            name: TOOL_NAME,
            description: TOOL_DESCRIPTION,
            inputSchema: {
              json: {
                type: "object",
                properties: {
                  description: {
                    type: "string",
                    description:
                      "Brief description of the letter or mailing content.",
                  },
                  rating: {
                    type: "string",
                    description:
                      "The eligibility rating for AdMail, as defined by ourselves, use with our prompt",
                    enum: [Rating.BUSINESS, Rating.UNSURE, Rating.ADVERTISING],
                  },
                  reason: {
                    type: "string",
                    description:
                      "Bullet pointed explaination of letter eligibility for Admail",
                  },
                  advice: {
                    type: "string",
                    description:
                      "Actionable bullet points to convert the letter to Admail, if applicable.",
                  },
                },
                required: ["description", "rating", "reason", "advice"],
              },
            },
          },
        },
      ],
      toolChoice: { tool: { name: TOOL_NAME } },
    };
  }

  async logPromptDetailsToS3({
    fileName,
    guardrailAssessment,
    promptInput,
    promptOutput,
  }: LogPromptDetailsInput): Promise<void> {
    const { loggingS3AccountId, loggingS3Bucket, loggingS3KeyPrefix } =
      this.config;

    if (!loggingS3Bucket || !loggingS3KeyPrefix || !loggingS3AccountId) {
      // eslint-disable-next-line no-console
      console.warn(ERROR_MESSAGES.S3_LOGGING_NOT_CONFIGURED);
      return;
    }

    const dateTimeNow = formatDateTimeForKey(new Date());
    const s3Key = `${loggingS3KeyPrefix}${dateTimeNow}|~${fileName}|~.json`;

    const logData = {
      prompt_input: promptInput,
      prompt_output: promptOutput,
      guardrail_assessment: guardrailAssessment,
      model: this.config.modelId,
      inference_parameters: {
        temperature: this.config.temperature,
        top_p: this.config.topP,
        max_tokens: this.config.maxTokens,
      },
      date_time: dateTimeNow,
    };

    try {
      await this.s3Client.send(
        new PutObjectCommand({
          Bucket: loggingS3Bucket,
          Key: s3Key,
          Body: JSON.stringify(logData, null, 4),
          ContentType: "application/json",
          ExpectedBucketOwner: loggingS3AccountId,
        }),
      );
    } catch (error) {
      // eslint-disable-next-line no-console
      console.error("Error logging to S3:", error);
    }
  }
}

export function createBedrockService(): BedrockService {
  // Imported lazily so the module can be imported in tests without
  // triggering real AWS client construction at module load time.
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  const { loadConfig } = require("./config") as typeof import("./config");
  const config = loadConfig();
  return new BedrockService(
    config,
    new BedrockRuntimeClient({ region: config.region }),
    new S3Client({}),
  );
}
